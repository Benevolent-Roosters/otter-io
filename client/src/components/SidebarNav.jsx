import React from 'react';
import { connect } from 'react-redux';
import { setCurrentBoard, toggleDrawer, toggleCreateBoard, getPanelsByBoard, getTicketsByPanel, setCurrentPanel, setPanels, setTickets, emptyPanels, emptyTickets  } from '../redux/actionCreators';
import { GridList, GridTile, Drawer } from 'material-ui';
import { Thumbnail, Image, Button } from 'react-bootstrap';
import moment from 'moment';
import { Link } from 'react-router-dom';

let counter = 0; 

let getRandomColor = function() {
  var letters = '0123456789ABCDEF';
  var color = '#';
  for (var i = 0; i < 6; i++) {
    color += letters[Math.floor(Math.random() * 16)];
  }
  return color;
};

const styles = {
  root: {
    display: 'flex',
    flexWrap: 'wrap',
    justifyContent: 'space-around',
    color: 'blue'
  },
  gridList: {
    marginTop: '25px',
    marginBottom: '25px'
  },

  tiles: {
    backgroundColor: getRandomColor()
  },

  profileImage: {
    display: 'block',
    margin: '0 auto', 
    marginTop: '15px', 
    marginLeft: '15px', 
    marginRight: '15px',
    backgroundColor: 'rgba(0, 0, 0, 0)',
    border: '0px',
    borderRadius: '0px'
  },

  createBoardImage: {
    display: 'block',
    height: 'auto',
    maxWidth: '100%',
    marginTop: '15px',
    marginRight: 'auto',
    marginLeft: 'auto',
    position: 'absolute',
    bottom: '50px',
    left: '64.5px',
    backgroundColor: 'rgba(0, 0, 0, 0)',
    border: '0px',
    borderRadius: '0px'
  },

  createBoardButton: {
    display: 'block',
    margin: '0 auto'
  }
};

const SidebarNavigation = (props) => {

  let findCurrentPanel = (panels) => {
    let dueDates = panels.map(panel => panel.due_date.slice(0, 10));
    let closest = 0;
    for (let i = 0; i < dueDates.length; i++) {
      if (moment().diff(dueDates[i]) < 0 && moment().diff(dueDates[i]) > moment().diff(dueDates[closest])) {
        closest = i;
      }
    }
    return closest;
  };

  let onBoardClick = (boardid) => {
    props.handleSetPanels(boardid, function(panels) {
      if (panels.length !== 0) {
        props.handleEmptyTickets();
        for (let panel of panels) {      
          props.handleSetTickets(panel.id);
        }
        let closestIndex = findCurrentPanel(panels);
        props.handleSetCurrentPanel(panels[closestIndex]);    
      } else {
        props.handleEmptyPanels();
        props.handleEmptyTickets();
      }
    });
  };

  return (
    <div>
      <Drawer containerStyle={{backgroundColor: '#ffffff'}} docked={false} open={props.drawerToggled} onRequestChange={(open) => props.handleToggledDrawer(open)}>
        <Thumbnail href="#" style={styles.profileImage} src={require('../images/business-person-silhouette-wearing-tie.png')} />

        <h3 style={{textAlign: 'center', marginBottom: '15px'}}>@dsc03</h3>

        <Button style={styles.createBoardButton} bsSize='large' bsStyle='primary' onClick={(open) => {props.handleToggledDrawer(open); props.handleCreateBoardRendered();}}>Create Board</Button>

        <GridList style={styles.gridList} cols={1} padding={15}>
          {props.boards.map(board => 
            <Link to={`/boards/${board.id}`}>
              <GridTile style={{marginLeft: '15px', marginRight: '15px', backgroundColor: getRandomColor()}} 
                key={board.id} title={board.board_name} 
                onClick={() => { onBoardClick(board.id); props.handleSetCurrentBoard(board); props.handleToggledDrawer();}}> 
              </GridTile>
            </Link>
            )
          }
        </GridList>

      </Drawer>
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    'gitHandle': state.rootReducer.user.github_handle,
    'profilePicture': state.rootReducer.user.profile_photo,
    'boards': state.rootReducer.boards,
    'currentBoard': state.rootReducer.currentBoard,
    'drawerToggled': state.rootReducer.drawerToggled
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    handleSetCurrentBoard(boardClicked) {
      dispatch(setCurrentBoard(boardClicked));
    }, 
    handleToggledDrawer(open) {
      dispatch(toggleDrawer(open));
    },
    handleCreateBoardRendered() {
      dispatch(toggleCreateBoard());
    },
    handleSetPanels(boardid, callback) {
      dispatch(getPanelsByBoard(boardid, callback));
    },
    handleSetTickets(panelId) {
      dispatch(getTicketsByPanel(panelId));
    },
    handleSetCurrentPanel(panel) {
      dispatch(setCurrentPanel(panel));
    },
    handleEmptyPanels() {
      dispatch(emptyPanels());
    },
    handleEmptyTickets() {
      dispatch(emptyTickets());
    }
  };
};

export var UnwrappedSidebar = SidebarNavigation;
export default connect(mapStateToProps, mapDispatchToProps)(SidebarNavigation);

// {props.boards.map((board) => {
//   return <GridTile key={require('../images/menu.png')} title={'Cheese Pizza'}></GridTile>;
// })}