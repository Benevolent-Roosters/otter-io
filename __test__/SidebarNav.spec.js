import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, mount} from 'enzyme';
import renderer from 'react-test-renderer';
import SidebarNav, { UnwrappedSidebar } from '../client/src/components/SidebarNav.jsx';
import { Provider } from 'react-redux';
import rootReducer from '../client/src/redux/reducers.js';

describe('<SidebarNav />', () => {
  let store = {
    gitHandle: 'something',
    profilePicture: 'someUrl',
    boards: [],
    currentBoard: {
      boardId: 0,
      teamName: '',
      repoName: '',
      owner: '',
      repoUrl: ''
    }
  };
  it ('should render the unconnected component correctly', () => {
    const wrapper = shallow(<UnwrappedSidebar />);
    expect(wrapper).toMatchSnapshot();
  });
  it ('should render the connected component correctly', () => {
    const wrapper = shallow(<Provider store={store}><SidebarNav /></Provider>);
    expect(wrapper).toMatchSnapshot();
  });
})