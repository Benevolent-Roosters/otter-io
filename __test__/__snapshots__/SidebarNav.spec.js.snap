// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SidebarNav /> should render the connected component correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Connect(SidebarNavigation) />,
  "nodes": Array [
    <Connect(SidebarNavigation) />,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Provider
        store={
                Object {
                        "boards": Array [],
                        "currentBoard": Object {
                          "boardId": 0,
                          "owner": "",
                          "repoName": "",
                          "repoUrl": "",
                          "teamName": "",
                        },
                        "gitHandle": "something",
                        "profilePicture": "someUrl",
                      }
        }
>
        <Connect(SidebarNavigation) />
</Provider>,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Provider {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "children": <Connect(SidebarNavigation) />,
          "store": Object {
            "boards": Array [],
            "currentBoard": Object {
              "boardId": 0,
              "owner": "",
              "repoName": "",
              "repoUrl": "",
              "teamName": "",
            },
            "gitHandle": "something",
            "profilePicture": "someUrl",
          },
        },
        "refs": Object {},
        "state": null,
        "store": Object {
          "boards": Array [],
          "currentBoard": Object {
            "boardId": 0,
            "owner": "",
            "repoName": "",
            "repoUrl": "",
            "teamName": "",
          },
          "gitHandle": "something",
          "profilePicture": "someUrl",
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Connect(SidebarNavigation) />,
        "_debugID": 4,
        "_renderedOutput": <Connect(SidebarNavigation) />,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Provider
    store={
        Object {
            "boards": Array [],
            "currentBoard": Object {
              "boardId": 0,
              "owner": "",
              "repoName": "",
              "repoUrl": "",
              "teamName": "",
            },
            "gitHandle": "something",
            "profilePicture": "someUrl",
          }
    }
>
    <Connect(SidebarNavigation) />
</Provider>,
}
`;

exports[`<SidebarNav /> should render the unconnected component correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    This is SidebarNavigation
</div>,
  "nodes": Array [
    <div>
      This is SidebarNavigation
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <SidebarNavigation />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {},
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          This is SidebarNavigation
</div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          This is SidebarNavigation
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <SidebarNavigation />,
}
`;
